File: repos/openrappter/typescript/src/security/secret-keys.ts
/**
 * One answer to "is this field name a secret?".
 *
 * There were two, and each missed what the other caught:
 *
 *   config display    token password key secret apikey api_key
 *   gateway logging   token password secret credential authorization
 *
 * So `apiKey`, `privateKey`, `signingKey` and `sessionKey` were written to the
 * structured log in the clear, while `credential`, `credentials` and
 * `authorization` were printed in the clear by `config show`. Two lists, two
 * different holes, neither visible from the other file.
 *
 * Matching is on word boundaries rather than substrings, so `apiKey` and
 * `private_key` are caught while `monkey` and `keyword` are not — the previous
 * config list redacted both of those by accident, which is harmless for display
 * but would blank useful fields if the same list were used for logs.
 */

/** Whole words that make a field a secret. */
const SECRET_WORDS = new Set([
  'apikey', 'auth', 'authorization', 'bearer', 'credential', 'credentials',
  'cookie', 'jwt', 'passphrase', 'passwd', 'password', 'pat', 'pem', 'secret',
  'secrets', 'signature', 'token', 'tokens',
]);

/**
 * Words that make a trailing `key` secret.
 *
 * `key` on its own is too blunt in both directions. Treating it as a whole word
 * blanked `keyCount`, `keyId` and `publicKey` — this file's own commit claimed
 * `keyCount` stayed readable, and it did not, which an outside review caught.
 * Dropping it entirely would let `apiKey` and `privateKey` through, which is
 * the bug the word was added for.
 *
 * So `key` counts when it is what the field *is* — the last word, qualified by
 * something that makes it sensitive — and not when it merely appears.
 */
const SECRET_KEY_QUALIFIERS = new Set([
  'access', 'api', 'app', 'auth', 'client', 'encryption', 'master', 'private',
  'secret', 'session', 'signing', 'ssh', 'token',
]);

/** Fragments that are unambiguous even when glued to other text. */
const SECRET_FRAGMENTS = [
  'apikey', 'api_key', 'authorization', 'credential', 'passphrase',
  'password', 'secret', 'token',
];

/** Split `apiKey`, `api_key`, `api-key` and `API KEY` into their words. */
function splitWords(key: string): string[] {
  return key
    .replace(/([a-z0-9])([A-Z])/g, '$1 $2')
    .replace(/([A-Z]+)([A-Z][a-z])/g, '$1 $2')
    .split(/[^A-Za-z0-9]+/)
    .map(part => part.toLowerCase())
    .filter(Boolean);
}

/**
 * True when one word makes a field a secret, singular or plural.
 *
 * `cookies` is exactly as sensitive as `cookie`. Plurals looked handled
 * because `tokens`, `secrets` and `credentials` are all caught — but they are
 * caught by the *fragment* pass, which is substring-based, not by being
 * spelled out above. So a word survived pluralisation only if it also happened
 * to be a fragment. `cookie`, `jwt` and `signature` are words only, and their
 * plurals fell through every branch into the clear.
 */
function isSecretWord(word: string): boolean {
  if (SECRET_WORDS.has(word)) return true;
  return word.endsWith('s') && SECRET_WORDS.has(word.slice(0, -1));
}

export function isSecretKey(key: string): boolean {
  const words = splitWords(key);
  if (words.some(word => isSecretWord(word))) return true;

  const last = words[words.length - 1];
  if (last === 'key' || last === 'keys') {
    // A field named exactly `key` is a value, not a label for one.
    if (words.length === 1) return true;
    if (words.slice(0, -1).some(word => SECRET_KEY_QUALIFIERS.has(word))) return true;
  }

  const lowered = key.toLowerCase();
  return SECRET_FRAGMENTS.some(fragment => lowered.includes(fragment));
}

File: repos/openrappter/python/openrappter/security/secret_keys.py
"""One answer to "is this field name a secret?".

The TypeScript runtime had two, and each missed what the other caught, so
`apiKey`, `privateKey`, `signingKey` and `sessionKey` were written into the
structured log in the clear. That was fixed in openrappter#76.

This body had the same logging pattern, character for character:

    token|password|secret|credential|authorization

so it leaked the same five names. Fixing one runtime and leaving the other is
how the two stop meaning the same thing.

Matching is on word boundaries rather than substrings, so `apiKey` and
`private_key` are caught while `monkey` and `keyword` are not — blanking a
field named `keyCount` would cost the debuggability the log exists for.

Kept deliberately in step with typescript/src/security/secret-keys.ts; a test
asserts the two agree on a shared table of names.
"""

from __future__ import annotations

import re

#: Whole words that make a field a secret.
_SECRET_WORDS = frozenset({
    "apikey", "auth", "authorization", "bearer", "credential", "credentials",
    "cookie", "jwt", "passphrase", "passwd", "password", "pat", "pem", "secret",
    "secrets", "signature", "token", "tokens",
})

#: Words that make a trailing ``key`` secret.
#:
#: ``key`` on its own is too blunt in both directions. As a whole word it
#: blanked ``keyCount``, ``keyId`` and ``publicKey``; dropping it would let
#: ``apiKey`` and ``privateKey`` through. So it counts when it is what the field
#: *is* — the last word, qualified by something that makes it sensitive.
_SECRET_KEY_QUALIFIERS = frozenset({
    "access", "api", "app", "auth", "client", "encryption", "master", "private",
    "secret", "session", "signing", "ssh", "token",
})

#: Fragments unambiguous even when glued to other text.
_SECRET_FRAGMENTS = (
    "apikey", "api_key", "authorization", "credential", "passphrase",
    "password", "secret", "token",
)

_CAMEL_BOUNDARY = re.compile(r"([a-z0-9])([A-Z])")
_ACRONYM_BOUNDARY = re.compile(r"([A-Z]+)([A-Z][a-z])")
_NON_ALNUM = re.compile(r"[^A-Za-z0-9]+")


def _split_words(key: str) -> list:
    """Split apiKey, api_key, api-key and API KEY into their words."""
    spaced = _CAMEL_BOUNDARY.sub(r"\1 \2", key)
    spaced = _ACRONYM_BOUNDARY.sub(r"\1 \2", spaced)
    return [part.lower() for part in _NON_ALNUM.split(spaced) if part]


def _is_secret_word(word: str) -> bool:
    """True when one word makes a field a secret, singular or plural.

    ``cookies`` is exactly as sensitive as ``cookie``. Plurals looked handled
    because ``tokens``, ``secrets`` and ``credentials`` are all caught — but
    they are caught by the *fragment* pass, which is substring-based, not by
    being spelled out above. So a word survived pluralisation only if it also
    happened to be a fragment. ``cookie``, ``jwt`` and ``signature`` are words
    only, and their plurals fell through every branch into the clear.
    """
    if word in _SECRET_WORDS:
        return True
    return word.endswith("s") and word[:-1] in _SECRET_WORDS


def is_secret_key(key: str) -> bool:
    """True when a field of this name must never be logged verbatim."""
    words = _split_words(key)
    if any(_is_secret_word(word) for word in words):
        return True

    if words and words[-1] in ("key", "keys"):
        # A field named exactly ``key`` is a value, not a label for one.
        if len(words) == 1:
            return True
        if any(word in _SECRET_KEY_QUALIFIERS for word in words[:-1]):
            return True

    lowered = key.lower()
    return any(fragment in lowered for fragment in _SECRET_FRAGMENTS)
